### PR TITLE
Extended page locking features, introduced PageLockedND arrays and removed ExchangeBuffer.

### PR DESCRIPTION
**Pull request by m4rs-mt**
This PR removes the `ExchangeBuffer` class in favor of newly implemented `PageLockedArrayND` types based on the `PageLockScope` (added in PR #480). These multi-dimensional array wrappers automatically pin and lock .Net arrays in memory to improve transfer speed.

This PR also adds a `PageLockMode` enumeration to automatically enable (and even force) page locking of implicitly instantiated memory buffers during transfer operations.

At the same time, this PR also addresses community feedback regarding the inconsistent naming of `GetAsNDArray` vs. `ArrayViewND` names by renaming `GetAsNDArray` to `GetAsArrayND` methods.

Sample:
```c#
// Specify automatic page locking during transfer operations
var context = Context.Create(builder => ... builder.PageLocking(PageLockingMode.Auto));
...

// Rely on disabled (default) or automatic page locking behavior
var array = buffer.GetAsArray1D();
// vs. an explicitly page locked version.
var pagedArray = buffer.GetAsPagedLockedArray1D();

// Allocate a page locked .Net array
var array = accelerator.AllocatePageLockedArray1D<long>(...);
```

Note that there are no extension methods for copying to/from `PageLockedArrayND` instances for all `ArrayViewND<Type, Stride>` specializations. This is because some transfer operations involve transposes, which in turn involve temporary transpose buffers.

### CODE REVIEW

File: Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs [hunk 1]
```diff
@@ -32,25 +32,38 @@ public sealed class CudaPageLockScope<T> : PageLockScope<T>
...
-            if (accelerator.Device.SupportsUsingHostPointerForRegisteredMemory)
+
+            // Perform the memory registration.
+            var cudaError = CurrentAPI.MemHostRegister(
+                hostPtr,
+                new IntPtr(LengthInBytes),
+                flags);
+
+            // Check whether we have already registered this memory pointer.
+            Locked = cudaError != CudaError.CUDA_ERROR_HOST_MEMORY_ALREADY_REGISTERED;
```
Comment (MoFtZ) [inline Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs:60]: ISSUE: When the memory address is already registered by another scope, there is no guarantee that the other scope will keep the registration longer than this scope.

**Review by MoFtZ — CHANGES_REQUESTED** [review]

File: Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs (hunk 1 above)
Comment (m4rs-mt) [reply Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs:60]: @MoFtZ Yes, you are right 🤓 I think we should revert these changes.

Comment (MoFtZ) [reply Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs:60]: Maybe we could introduce a reference-counting system?

Comment (m4rs-mt) [reply Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs:60]: @MoFtZ I'm not sure it's worth introducing reference counting in this context.

**Review by MoFtZ — APPROVED** [review]

**Pull request merged**